src/vs/workbench/services/languageDetection/browser/languageDetectionWorkerServiceImpl.ts

```diff
@@ -0,0 +1,287 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Disposable, DisposableStore, dispose, IDisposable, toDisposable } from 'vs/base/common/lifecycle';
+import { ILanguageDetectionService } from 'vs/workbench/services/languageDetection/common/languageDetectionWorkerService';
+import { FileAccess } from 'vs/base/common/network';
+import { IWorkbenchEnvironmentService } from 'vs/workbench/services/environment/common/environmentService';
+import { IConfigurationService } from 'vs/platform/configuration/common/configuration';
+import { IModeService } from 'vs/editor/common/services/modeService';
+import { URI } from 'vs/base/common/uri';
+import { isWeb } from 'vs/base/common/platform';
+import { registerSingleton } from 'vs/platform/instantiation/common/extensions';
+import { LanguageDetectionSimpleWorker } from 'vs/workbench/services/languageDetection/browser/languageDetectionSimpleWorker';
+import { DefaultWorkerFactory } from 'vs/base/worker/defaultWorkerFactory';
+import { IModelService } from 'vs/editor/common/services/modelService';
+import { IntervalTimer } from 'vs/base/common/async';
+import { SimpleWorkerClient } from 'vs/base/common/worker/simpleWorker';
+import { ITelemetryService } from 'vs/platform/telemetry/common/telemetry';
+
+export class LanguageDetectionService extends Disposable implements ILanguageDetectionService {
+	static readonly enablementSettingKey = 'workbench.editor.untitled.experimentalLanguageDetection';
+
+	_serviceBrand: undefined;
+
+	private _languageDetectionWorkerClient: LanguageDetectionWorkerClient;
+
+	constructor(
+		@IWorkbenchEnvironmentService private readonly _environmentService: IWorkbenchEnvironmentService,
+		@IModeService private readonly _modeService: IModeService,
+		@IConfigurationService private readonly _configurationService: IConfigurationService,
+		@IModelService modelService: IModelService,
+		@ITelemetryService telemetryService: ITelemetryService,
+	) {
+		super();
+
+		this._languageDetectionWorkerClient = new LanguageDetectionWorkerClient(
+			modelService,
+			telemetryService,
+			'languageDetectionWorkerService',
+			// TODO: See if it's possible to bundle vscode-languagedetection
+			this._environmentService.isBuilt && !isWeb
```

Comment:
This would probably read easier if the common prefix up until `vscode-languagedetection` was extracted to a variable.